In [1]:
from sklearn.datasets import load_iris
from sklearn import linear_model
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
iris = load_iris()
data1 = pd.DataFrame(data= np.c_[iris['data'], iris['target']],
                     columns= iris['feature_names'] + ['target'])

In [3]:
data1

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
..      ...  
145     2.0  
146     2.0  
147     2.0  
148     2.0  
149     2.0  

[150 rows x 5 columns]

X is always the INPUT variable (independent variable)
Y is always the OUTPUT variable (dependent variable, variable we're getting a prediction for)

In [4]:
X = data1[['petal length (cm)']]
X

petal length (cm)
0                  1.4
1                  1.4
2                  1.3
3                  1.5
4                  1.4
..                 ...
145                5.2
146                5.0
147                5.2
148                5.4
149                5.1

[150 rows x 1 columns]

In [5]:
Y = data1['petal width (cm)']
Y

0      0.2
1      0.2
2      0.2
3      0.2
4      0.2
      ... 
145    2.3
146    1.9
147    2.0
148    2.3
149    1.8
Name: petal width (cm), Length: 150, dtype: float64

In [6]:
#Step 1: Define the model
lm = linear_model.LinearRegression()

#Step 2: Fitting the model
lm.fit(X,Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [7]:
print(lm.intercept_)

-0.3630755213190291


In [8]:
print(lm.coef_)

[0.41575542]


In [9]:
Y = -0.363 + 0.416*(X)

In [10]:
def model(X):
    return -0.363 + 0.416*(X)

model(5.8)

2.0498

In [11]:
data1

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
..      ...  
145     2.0  
146     2.0  
147     2.0  
148     2.0  
149     2.0  

[150 rows x 5 columns]

In [12]:
#Input Variable (dataframe)
Input = data1[['petal width (cm)']]

#Output Variable (series)
Output = data1['sepal length (cm)']

#Defining the model
lm = linear_model.LinearRegression()

#Fit the model
lm.fit(Input, Output)

#Print the intercept
print('Intercept: ' + str(lm.intercept_))

#Print the coefficient
print('Coefficient ' + str(lm.coef_))

print('\n')

#Print the full model
print('Sepal Length (CM) = ' + str(round(lm.intercept_,3)) + ' + ' + str(round(lm.coef_[0],3)) + '*(Petal Width (CM))')


Intercept: 4.777629415601567
Coefficient [0.88858025]


Sepal Length (CM) = 4.778 + 0.889*(Petal Width (CM))


In [21]:
data1[:120]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
115                6.4               3.2                5.3               2.3   
116                6.5               3.0                5.5               1.8   
117                7.7               3.8                6.7               2.2   
118                7.7               2.6                6.9               2.3   
119                6.0               2.2                5.0               1.5   

     target  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
..      ...  
115     2.0  
116     2.0  
117     2.0  
118     2.0  
119     2.0  

[120 rows x 5 columns]

In [23]:
data1[120:]

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
120                6.9               3.2                5.7               2.3   
121                5.6               2.8                4.9               2.0   
122                7.7               2.8                6.7               2.0   
123                6.3               2.7                4.9               1.8   
124                6.7               3.3                5.7               2.1   
125                7.2               3.2                6.0               1.8   
126                6.2               2.8                4.8               1.8   
127                6.1               3.0                4.9               1.8   
128                6.4               2.8                5.6               2.1   
129                7.2               3.0                5.8               1.6   
130                7.4               2.8                6.1               1.9   
131                7.9               3.8                6.4               2.0   
132                6.4               2.8                5.6               2.2   
133                6.3               2.8                5.1               1.5   
134                6.1               2.6                5.6               1.4   
135                7.7               3.0                6.1               2.3   
136                6.3               3.4                5.6               2.4   
137                6.4               3.1                5.5               1.8   
138                6.0               3.0                4.8               1.8   
139                6.9               3.1                5.4               2.1   
140                6.7               3.1                5.6               2.4   
141                6.9               3.1                5.1               2.3   
142                5.8               2.7                5.1               1.9   
143                6.8               3.2                5.9               2.3   
144                6.7               3.3                5.7               2.5   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
120     2.0  
121     2.0  
122     2.0  
123     2.0  
124     2.0  
125     2.0  
126     2.0  
127     2.0  
128     2.0  
129     2.0  
130     2.0  
131     2.0  
132     2.0  
133     2.0  
134     2.0  
135     2.0  
136     2.0  
137     2.0  
138     2.0  
139     2.0  
140     2.0  
141     2.0  
142     2.0  
143     2.0  
144     2.0  
145     2.0  
146     2.0  
147     2.0  
148     2.0  
149     2.0

In [19]:
#Select first 3 columns
input_vars = list(data1.columns[0:3])

#Input (to be a dataframe structure)
Input = data1[input_vars]

#Output (to be a series structure)
Output = data1['petal width (cm)']

#Define the model type
lm = linear_model.LinearRegression()

#Fit model
lm.fit(Input, Output)

#Print Intercept
print(lm.intercept_)

#Print Coefficient
print(lm.coef_)

-0.2403073891122598
[-0.20726607  0.22282854  0.52408311]


Petal Width = -0.240 + -0.207*(Sepal Length) + 0.222*(Sepal Width) + 0.524*(Petal Length)

# Model Accuracy

In [20]:
lm.score(Input,Output)

0.9378502736046809

# Train/Test Split

In [28]:
# Fitting the models

X_train, X_test, y_train, y_test = train_test_split(Input, Output, test_size=0.2, random_state=0)
lm.fit(X_train, y_train) #fit the model
print(lm.score(X_train, y_train)) #how it performs on the data it was trained on


0.9463080776103777


In [29]:
X_train

sepal length (cm)  sepal width (cm)  petal length (cm)
137                6.4               3.1                5.5
84                 5.4               3.0                4.5
27                 5.2               3.5                1.5
127                6.1               3.0                4.9
132                6.4               2.8                5.6
..                 ...               ...                ...
9                  4.9               3.1                1.5
103                6.3               2.9                5.6
67                 5.8               2.7                4.1
117                7.7               3.8                6.7
47                 4.6               3.2                1.4

[120 rows x 3 columns]

In [30]:
print(lm.score(X_test, y_test)) #how it performs on data we held out

0.891338738620735


In [32]:
y_test

114    2.4
62     1.0
33     0.2
107    1.8
7      0.2
100    2.5
40     0.3
86     1.5
76     1.4
71     1.3
134    1.4
51     1.5
73     1.2
54     1.5
63     1.4
37     0.1
78     1.5
90     1.2
45     0.3
16     0.4
121    2.0
66     1.5
24     0.2
8      0.2
126    1.8
22     0.2
44     0.4
97     1.3
93     1.0
26     0.4
Name: petal width (cm), dtype: float64